File: Source/Aladdin.Net/RemoteHost.cs
using System;
using System.Net;
using System.Net.Sockets;

namespace Aladdin.Net
{
    ///////////////////////////////////////////////////////////////////////////
    // Удаленный хост
    ///////////////////////////////////////////////////////////////////////////
    public class RemoteHost
    {
        // список допустимых адресов
        private IPAddress[] addressList; 

        // конструктор
        public RemoteHost(IPAddress[] addressList) { this.addressList = addressList; }

        // конструктор
        public RemoteHost(IPHostEntry hostEntry)
        {
            // при наличии списка адресов 
            if (hostEntry.AddressList != null && hostEntry.AddressList.Length > 0)
            {
                // указать список адресов
                this.addressList = hostEntry.AddressList; 
            }
            // определить адреса
            else { this.addressList = Dns.GetHostAddresses(hostEntry.HostName); }
        }
        // конструктор
        public RemoteHost(string hostName)
        {
            // определить адреса
            this.addressList = Dns.GetHostAddresses(hostName); 
        }
        // список допустимых адресов
        public IPAddress[] AddressList { get { return addressList; }}

        // создать соединение
        public Socket Connect(int port, SocketType socketType, ProtocolType protocolType)
        {
            // указать начальные условия
            Exception exception = null;

            // для всех полученных адресов 
            foreach (IPAddress address in addressList)
            try {
                // для IP6-адреса
                if (address.AddressFamily == AddressFamily.InterNetworkV6) 
                {
                    // для проверить поддержку IP6-адреса
                    if (!Socket.OSSupportsIPv6) continue; 
                }
                // указать способ создания сокета
                Socket socket = new Socket(address.AddressFamily, socketType, protocolType); 

                // выполнить соединение
                try { socket.Connect(address, port); return socket; } 
                
                // обработать возможную ошибку
                catch { socket.Close(); socket = null; throw; }
            }
            // обработать возможное исключение
            catch (Exception e) { if (exception == null) exception = e; } throw exception;
        }
    }
}

File: Source/Aladdin.Net.Java/src/aladdin/net/RemoteHost.java
package aladdin.net;
import java.net.*;
import java.io.*;

///////////////////////////////////////////////////////////////////////////
// Удаленный хост
///////////////////////////////////////////////////////////////////////////
public class RemoteHost
{
    // список допустимых адресов
    private final InetAddress[] addressList; 

    // конструктор
    public RemoteHost(InetAddress[] addressList) { this.addressList = addressList; }

    // конструктор
    public RemoteHost(String hostName) throws UnknownHostException
    {
        // определить адреса
        this.addressList = InetAddress.getAllByName(hostName); 
    }
    // список допустимых адресов
    public final InetAddress[] addressList() { return addressList; } 
    
    ///////////////////////////////////////////////////////////////////////////
    // TCP
    ///////////////////////////////////////////////////////////////////////////
    public final Socket connect(int port) throws IOException
    {
        // для всех полученных адресов 
        IOException exception = null; for (InetAddress address : addressList)
        try {
            // указать способ создания сокета
            return new Socket(address, port); 
        }
        // обработать возможное исключение
        catch (IOException e) { if (exception == null) exception = e; } throw exception; 
    }
    ///////////////////////////////////////////////////////////////////////////
    // UDP
    ///////////////////////////////////////////////////////////////////////////
    public final DatagramSocket connectDatagram(int port) throws IOException
    {
        // открыть сокет
        DatagramSocket socket = new DatagramSocket(); 

        // для всех полученных адресов 
        Throwable exception = null; for (InetAddress address : addressList)
        try {
            // выполнить соединение
            socket.connect(address, port); return socket; 
        }
        // обработать возможное исключение
        catch (Throwable e) { if (exception == null) exception = e; } 
        
        // проверить отсутствие ошибок
        if (exception != null) socket.close(); throw new IOException(exception);
    }
}
